### PR TITLE
Annotated SyntaxList

### PR DESCRIPTION
**Pull request by jaredpar**

### CODE REVIEW

**Comment by jaredpar** [comment]
@roslyn-compiler PTAL

File: src/Compilers/Core/Portable/Syntax/SyntaxList`1.cs [hunk 1]
```diff
@@ -17,9 +19,9 @@ namespace Microsoft.CodeAnalysis
     public readonly partial struct SyntaxList<TNode> : IReadOnlyList<TNode>, IEquatable<SyntaxList<TNode>>
         where TNode : SyntaxNode
     {
-        private readonly SyntaxNode _node;
+        private readonly SyntaxNode? _node;
 
-        internal SyntaxList(SyntaxNode node)
+        internal SyntaxList(SyntaxNode? node)
```
Comment (cston) [inline src/Compilers/Core/Portable/Syntax/SyntaxList`1.cs]: >SyntaxNode? [](start = 28, length = 11)

When is this `null`?

Comment (jaredpar) [reply src/Compilers/Core/Portable/Syntax/SyntaxList`1.cs]: Most obvious place is here:

```
        public SyntaxList(IEnumerable<TNode>? nodes)
            : this(CreateNode(nodes)!)
```

Which ... I should really remove the `!` from cause it's unnecessary.

File: src/Compilers/Core/Portable/Syntax/SyntaxList.SeparatedWithManyChildren.cs [hunk 2]
```diff
@@ -10,24 +12,24 @@ internal class SeparatedWithManyChildren : SyntaxList
...
+            internal override SyntaxNode? GetNodeSlot(int i)
             {
                 if ((i & 1) != 0)
                 {
                     //separator
                     return null;
                 }
 
-                return this.GetRedElement(ref _children[i >> 1].Value, i);
+                return this.GetRedElement(ref _children[i >> 1].Value!, i);
```
Comment (sharwell) [inline src/Compilers/Core/Portable/Syntax/SyntaxList.SeparatedWithManyChildren.cs]: ❔ Why this suppression?

File: src/Compilers/Core/Portable/Syntax/SyntaxList.SeparatedWithManyChildren.cs [hunk 3]
```diff
@@ -10,24 +12,24 @@ internal class SeparatedWithManyChildren : SyntaxList
         {
             private readonly ArrayElement<SyntaxNode>[] _children;
```
Comment (sharwell) [inline src/Compilers/Core/Portable/Syntax/SyntaxList.SeparatedWithManyChildren.cs]: 💡 Changing this to `ArrayElement<SyntaxNode?>[]` seems to be an improvement for this type

File: src/Compilers/Core/Portable/Syntax/SyntaxList.cs [hunk 4]
```diff
@@ -21,9 +23,9 @@ public override string Language
             }
         }
 
-        protected override SyntaxTree SyntaxTreeCore => this.Parent.SyntaxTree;
+        protected override SyntaxTree SyntaxTreeCore => this.Parent!.SyntaxTree;
```
Comment (sharwell) [inline src/Compilers/Core/Portable/Syntax/SyntaxList.cs]: ⚠️  This appears to violate the documented behavior of the base property (emphasis mine):

> Returns SyntaxTree that owns the node **or null if node does not belong to a SyntaxTree**

```suggestion
        protected override SyntaxTree SyntaxTreeCore => this.Parent?.SyntaxTree;
```

File: src/Compilers/Core/Portable/Syntax/SyntaxList.SeparatedWithManyChildren.cs (hunk 2 above)
Comment (jaredpar) [reply src/Compilers/Core/Portable/Syntax/SyntaxList.SeparatedWithManyChildren.cs]: The parameter here expects a nullable input but this is non-null

File: src/Compilers/Core/Portable/Syntax/SyntaxListBuilder.cs [hunk 5]
```diff
@@ -81,15 +83,15 @@ public void AddRange(SyntaxList<SyntaxNode> list)
...
+            if (this.Count + count > _nodes.Length)
             {
                 this.Grow(Count + count);
             }
 
             var dst = this.Count;
             for (int i = offset, limit = offset + count; i < limit; i++)
             {
-                _nodes[dst].Value = list.ItemInternal(i).Green;
+                _nodes[dst].Value = list.ItemInternal(i)!.Green;
```
Comment (sharwell) [inline src/Compilers/Core/Portable/Syntax/SyntaxListBuilder.cs]: 💡 Consider splitting the expression `list.ItemInternal(i)` to a local, then asserting that it is not null.

File: src/Compilers/Core/Portable/Syntax/SyntaxListBuilder.cs [hunk 6]
```diff
@@ -153,7 +155,7 @@ public bool Any(int kind)
         {
             for (int i = 0; i < Count; i++)
             {
-                if (_nodes[i].Value.RawKind == kind)
+                if (_nodes[i].Value!.RawKind == kind)
```
Comment (sharwell) [inline src/Compilers/Core/Portable/Syntax/SyntaxListBuilder.cs]: ❔ Is it possible for this to reach external code and still be null?

File: src/Compilers/Core/Portable/Syntax/SyntaxList.cs (hunk 4 above)
Comment (jaredpar) [reply src/Compilers/Core/Portable/Syntax/SyntaxList.cs]: It does violate the documented behavior but it is also the behavior. Safe bug to fix here.

File: src/Compilers/Core/Portable/Syntax/SyntaxListBuilderExtensions.cs [hunk 7]
```diff
@@ -42,7 +44,7 @@ public static SyntaxList<TNode> ToList<TNode>(this SyntaxListBuilder builder)
                 return default(SeparatedSyntaxList<TNode>);
             }
 
-            return new SeparatedSyntaxList<TNode>(new SyntaxNodeOrTokenList(builder.ToListNode().CreateRed(), 0));
+            return new SeparatedSyntaxList<TNode>(new SyntaxNodeOrTokenList(builder.ToListNode()!.CreateRed(), 0));
```
Comment (sharwell) [inline src/Compilers/Core/Portable/Syntax/SyntaxListBuilderExtensions.cs]: 💡 Consider refactoring this method:

```
var listNode = builder.ToListNode();
if (listNode is null)
{
  return null;
}

return new SeparatedSyntaxList<TNode>(new SyntaxNodeOrTokenList(listNode.CreateRed(), 0));
```

Applies to methods above too.

File: src/Compilers/Core/Portable/Syntax/SyntaxListBuilderExtensions.cs [hunk 8]
```diff
@@ -32,7 +34,7 @@ public static SyntaxList<TNode> ToList<TNode>(this SyntaxListBuilder builder)
                 return new SyntaxList<TNode>();
             }
 
-            return new SyntaxList<TNode>(builder.ToListNode().CreateRed());
+            return new SyntaxList<TNode>(builder.ToListNode()!.CreateRed());
         }
 
         public static SeparatedSyntaxList<TNode> ToSeparatedList<TNode>(this SyntaxListBuilder builder) where TNode : SyntaxNode
```
Comment (sharwell) [inline src/Compilers/Core/Portable/Syntax/SyntaxListBuilderExtensions.cs]: 💡 The implementation allows `builder` to be null. I accounted for this in my suggested refactoring.

File: src/Compilers/Core/Portable/Syntax/SyntaxList`1.cs [hunk 9]
```diff
@@ -106,9 +108,9 @@ public int Count
             }
```
Comment (sharwell) [inline src/Compilers/Core/Portable/Syntax/SyntaxList`1.cs]: 📝 The `ArgumentOutOfRangeException` should have `nameof(index)`

File: src/Compilers/Core/Portable/Syntax/SyntaxListBuilder.cs (hunk 6 above)
Comment (jaredpar) [reply src/Compilers/Core/Portable/Syntax/SyntaxListBuilder.cs]: Not that I'm aware of. It's an implementation detail of this type.

---
In reply to: [362963462](https://github.com/dotnet/roslyn/pull/40729#discussion_r362963462) [](ancestors = 362963462)

File: src/Compilers/Core/Portable/Syntax/SyntaxListBuilderExtensions.cs (hunk 7 above)
Comment (jaredpar) [reply src/Compilers/Core/Portable/Syntax/SyntaxListBuilderExtensions.cs]: That would change the return type of the method to be a nullable struct. That's not a trivial change here. 

---
In reply to: [362969063](https://github.com/dotnet/roslyn/pull/40729#discussion_r362969063) [](ancestors = 362969063)

Comment (jaredpar) [reply src/Compilers/Core/Portable/Syntax/SyntaxListBuilderExtensions.cs]: Did u mean `return default`?

---
In reply to: [362974659](https://github.com/dotnet/roslyn/pull/40729#discussion_r362974659) [](ancestors = 362974659,362969063)

Comment (sharwell) [reply src/Compilers/Core/Portable/Syntax/SyntaxListBuilderExtensions.cs]: Yes I meant `return default`

**Comment by 333fred** [comment]
>                 return _builder == null;

This implies `_builder` should be annotated. Why is it not?

---
Refers to: src/Compilers/Core/Portable/Syntax/SyntaxListBuilder`1.cs:30 in 8e507a5. [](commit_id = 8e507a5b21c1d11a326f8094f9d8328f14462a13, deletion_comment = False)

**Comment by cston** [comment]
>             }

Should this use `var listNode = builder?.ToListNode();` for consistency with extension methods below?

---
Refers to: src/Compilers/Core/Portable/Syntax/SyntaxListBuilderExtensions.cs:14 in 8e507a5. [](commit_id = 8e507a5b21c1d11a326f8094f9d8328f14462a13, deletion_comment = False)

**Comment by jaredpar** [comment]
>                 return _builder == null;

It's a good question. One of the tricky parts of nullability is how we treat reference fields inside `struct`. Do we annotate for the `default` case or the case where we assume it's always properly constructed. Our `struct` implementation also suggest that we struggle with this type of decision. So far I've been leaning towards "annotate for default". In this case I just missed it. Will update. 

Overall though I don't think this type of scenario is cut and dry. 

---
In reply to: [570725573](https://github.com/dotnet/roslyn/pull/40729#issuecomment-570725573) [](ancestors = 570725573)

---
Refers to: src/Compilers/Core/Portable/Syntax/SyntaxListBuilder`1.cs:30 in 8e507a5. [](commit_id = 8e507a5b21c1d11a326f8094f9d8328f14462a13, deletion_comment = False)

**Pull request merged**